File: billing_system/frontend/flutter_application/lib/models/product.dart
/// Product model — no GST, plain sale price.
class Product {
  final String id;
  final String name;
  final String unit;
  final double price;   // final sale price (no tax)
  final double mrp;
  final double stock;
  final String category;
  final String description;
  final String? imageUrl;

  const Product({
    required this.id,
    required this.name,
    required this.unit,
    required this.price,
    required this.mrp,
    required this.stock,
    required this.category,
    this.description = '',
    this.imageUrl,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id:          json['id'].toString(),
      name:        json['name'] as String,
      unit:        json['unit'] as String,
      price:       (json['price'] as num).toDouble(),
      mrp:         (json['mrp'] as num).toDouble(),
      stock:       (json['stock'] as num).toDouble(),
      category:    json['category'] as String? ?? '',
      description: json['description'] as String? ?? '',
      imageUrl:    json['image_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'unit': unit,
        'price': price,
        'mrp': mrp,
        'stock': stock,
        'category': category,
        'description': description,
        'image_url': imageUrl,
      };

  // ── Computed (GST not applicable — kept for UI compatibility) ────────────
  double get gst         => 0.0;           // no GST in this system
  double get priceWithGst => price;        // price is already the final price

  @override
  String toString() => 'Product($id, $name)';
}

File: billing_system/backend/models/product.py
"""
Product model – plain Python dataclass.
No ORM dependency so the migration path to SQLAlchemy is straightforward:
replace the dataclass with a db.Model subclass and keep every other file intact.
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class Product:
    id: str            # UUID from Supabase (e.g. "0ddb5e4f-d5ce-...")
    name: str
    unit: str          # e.g. "KG", "PCS", "LTR"
    price: float       # sale price (final price, no GST)
    mrp: float         # maximum retail price
    stock: float       # available stock quantity
    category: str      # e.g. "Grocery", "Dairy"
    description: Optional[str] = field(default="")
    image_url: Optional[str] = field(default=None)

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "unit": self.unit,
            "price": self.price,
            "mrp": self.mrp,
            "stock": self.stock,
            "category": self.category,
            "description": self.description,
            "image_url": self.image_url,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Product":
        return cls(
            id=str(data["id"]),
            name=data["name"],
            unit=data["unit"],
            price=float(data["price"]),
            mrp=float(data.get("mrp", data["price"])),
            stock=float(data["stock"]),
            category=data.get("category", "General"),
            description=data.get("description", ""),
            image_url=data.get("image_url"),
        )
